### DISCUSSION TITLE
OAuth support for Google monitoring prometheus backend

### DISCUSSION THREAD

**Discussion by sergei-koren** [grafana/grafana#40844] [Data Sources]
Hello,
Google has introduced Prometheus backend support for the Google Monitoring suite.
https://cloud.google.com/stackdriver/docs/managed-prometheus
Grafana dashboard is a vital part of any Prometheus compatible ecosystem.
As usual, every GCP API is protected by OAuth authentication, which currently is not supported by Grafana for the Prometheus backend.
I would appreciate Grafana to make a decision and begin either supporting Auth method on the Prometheus backend or by adding Promql support on the Google monitoring backend.

**Comment by vickyyyyyyy** [comment]
GCM has introduced PromQL query support since Sep 26 2022: https://cloud.google.com/release-notes#September_26_2022

**Reply by andresmgot** [reply]
Since this is now supported, we could add support for it in two different ways:

 - Add a PromQL editor in the Google Cloud Monitoring plugin.
 - Add Google Cloud authentication support for Prometheus.

From an engineering perspective, the second is much more viable. The existing Prometheus plugin is not designed to be used as a dependency so that means that it should go through a heavy refactor if we were to go for the first (or create a brand new editor, which would take as much and would have less functionality).

If we add support for gcloud auth in Prometheus (as done for SigV4 or Azure), then the implementation becomes trivial. To demonstrate this, I have created a Prometheus data source, skipping the built-in authentication methods and hard-coding an Authorization header (which I obtained using the Google Cloud Monitoring plugin):

![Screenshot from 2022-12-07 12-42-13](https://user-images.githubusercontent.com/4025665/206170732-2588f2d7-8c76-4552-b7d8-6d6c3440882e.png)

After that, I was able to use the PromQL query builder (with some issues, detailed below):

![Screenshot from 2022-12-07 12-43-59](https://user-images.githubusercontent.com/4025665/206171046-33b88f0a-92e1-46ff-8253-2a9d275f0134.png)

The documentation for the new PromQL API can be found here:

https://cloud.google.com/stackdriver/docs/managed-prometheus/query#api-prometheus

When building a query I found a couple of issues:

 - Searching for metrics won't work, even though all metrics are listed in the selector.
![Screenshot from 2022-12-07 12-46-04](https://user-images.githubusercontent.com/4025665/206171480-a84a2a55-d255-4e35-8f27-abe7de96ce53.png)
 - A similar issue appears when obtaining possible values for a filter. The filter options are listed properly but values are not. Failing with:
![Screenshot from 2022-12-07 12-50-01](https://user-images.githubusercontent.com/4025665/206172379-f16a953a-7bfa-44dc-81f5-a3bc9c4cf15b.png)

Both issues are known issues and the reason is detailed in the [API compatibility section](https://cloud.google.com/stackdriver/docs/managed-prometheus/query#http-api-details) of the documentation.

> The /api/v1/$label/values endpoint supports only the __name__ label. This limitation causes label_values($label) variable queries in Grafana to fail. Instead, you can use label_values($metric, $label). This type of query is recommended because it avoids fetching values for labels on metrics that are not relevant to the given dashboard.

**Reply by mhoran** [reply]
This would be great to see. I had experimented with this as well and it seems to work fine, with the exception of the known issues. The [frontend](https://github.com/GoogleCloudPlatform/prometheus-engine/tree/main/cmd/frontend) that one must run now is cumbersome and introduces unnecessary latency.

**Comment by DawPawel** [comment]
Heyo,
any update on this?

**Comment by simonblund** [comment]
Upvote on this!

**Comment by bwplotka** [comment]
Thanks for reporting. I proposed alternative proposal https://github.com/grafana/grafana/discussions/69180, but this one would solve the problem as well. 🤗  

We are happy to contribute this change as well if Grafana would accept it 👍🏽

**Comment by natellium** [comment]
Hey there! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

I will be closing this discussion, but if you’re still interested in this request being implemented please don’t hesitate to open a [feature request](https://github.com/grafana/grafana/issues/new?assignees=&labels=&projects=&template=1-feature_requests.md). Sorry for the inconvenience, and thanks for your contributions!

cc @grafana/partner-datasources

**Reply by mhoran** [reply]
I have opened a feature request here: https://github.com/grafana/grafana/issues/79410.
